dbms/src/Flash/Mpp/MPPTaskManager.cpp

```diff
@@ -27,13 +27,102 @@ namespace DB
 namespace FailPoints
 {
 extern const char random_task_manager_find_task_failure_failpoint[];
+extern const char pause_before_register_non_root_mpp_task[];
 } // namespace FailPoints
 
 MPPTaskManager::MPPTaskManager(MPPTaskSchedulerPtr scheduler_)
     : scheduler(std::move(scheduler_))
     , log(Logger::get("TaskManager"))
 {}
 
+MPPQueryTaskSetPtr MPPTaskManager::addMPPQueryTaskSet(UInt64 query_id)
+{
+    auto ptr = std::make_shared<MPPQueryTaskSet>();
+    mpp_query_map.insert({query_id, ptr});
+    GET_METRIC(tiflash_mpp_task_manager, type_mpp_query_count).Set(mpp_query_map.size());
+    return ptr;
+}
+
+void MPPTaskManager::removeMPPQueryTaskSet(UInt64 query_id, bool on_abort)
+{
+    scheduler->deleteQuery(query_id, *this, on_abort);
+    mpp_query_map.erase(query_id);
+    GET_METRIC(tiflash_mpp_task_manager, type_mpp_query_count).Set(mpp_query_map.size());
+}
+
+std::pair<MPPTunnelPtr, String> MPPTaskManager::findAsyncTunnel(const ::mpp::EstablishMPPConnectionRequest * request, EstablishCallData * call_data, grpc::CompletionQueue * cq)
+{
+    const auto & meta = request->sender_meta();
+    MPPTaskId id{meta.start_ts(), meta.task_id()};
+    Int64 sender_task_id = meta.task_id();
+    Int64 receiver_task_id = request->receiver_meta().task_id();
+
+    std::unique_lock lock(mu);
+    auto query_it = mpp_query_map.find(id.start_ts);
+    if (query_it == mpp_query_map.end())
+    {
+        if (!call_data->isWaitingTunnelState())
+        {
+            /// if call_data is in new_request state, then put it to waiting tunnel state
+            auto query_set = addMPPQueryTaskSet(id.start_ts);
+            auto & alarm = query_set->alarms[sender_task_id][receiver_task_id];
+            call_data->setToWaitingTunnelState();
+            alarm.Set(cq, Clock::now() + std::chrono::seconds(10), call_data);
+            return {nullptr, ""};
+        }
+        else
+        {
+            return {nullptr, fmt::format("Can't find task [{},{}] within 10 s.", id.start_ts, id.task_id)};
+        }
+    }
+    auto query_set = query_it->second;
+
+    if (!query_set->isInNormalState())
+    {
+        /// if the query is aborted, return true to stop waiting timeout.
+        LOG_WARNING(log, fmt::format("Query {} is aborted, all its tasks are invalid.", id.start_ts));
+        /// meet error
+        return {nullptr, query_it->second->error_message};
+    }
+    auto it = query_set->task_map.find(id);
+    if (it == query_set->task_map.end())
+    {
+        if (!call_data->isWaitingTunnelState())
+        {
+            /// task not found, use alarm
+            auto & alarm = query_set->alarms[sender_task_id][receiver_task_id];
```

Comment:
do we need null check for `alarms[sender_task_id]` and `alarms[sender_task_id][receiver_task_id]`